### PR TITLE
feat: add genfile pkg

### PR DESCRIPTION
**Pull request by katcipis**
Implements generate_file: https://github.com/mineiros-io/terramate/blob/main/docs/codegen/generate-file.md

This PR only adds the core logic via the genfile pkg, it doesn't integrate the pkg on terramate itself, so the feature won't be available yet. The next step is to integrate genfile in a similar fashion as genhcl is integrated on the core generate pkg.

Thanks to @kassianh for the initial work on this feature.

### CODE REVIEW

File: generate/genfile/genfile_test.go [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+	}
+	globals := func(builders ...hclwrite.BlockBuilder) *hclwrite.Block {
+		return hclwrite.BuildBlock("globals", builders...)
+	}
+	expr := func(name string, expr string) hclwrite.BlockBuilder {
+		return hclwrite.Expression(name, expr)
+	}
+	str := func(name string, val string) hclwrite.BlockBuilder {
+		return hclwrite.String(name, val)
+	}
```
Comment (i4ki) [inline generate/genfile/genfile_test.go:72]:
```suggestion
	hcldoc := hclwrite.BuildHCL
	labels := hclwrite.Labels
	expr := hclwrite.Expression
	str := hclwrite.String
	generateFile := func(builders ...hclwrite.BlockBuilder) *hclwrite.Block {
		return hclwrite.BuildBlock("generate_file", builders...)
	}
	globals := func(builders ...hclwrite.BlockBuilder) *hclwrite.Block {
		return hclwrite.BuildBlock("globals", builders...)
	}
```

just a simplification.

File: generate/genfile/genfile_test.go [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+						str("content", "${global.data}-${terramate.path}"),
+					),
+				},
+			},
+			want: []result{
+				{
+					name: "test",
+					file: genFile{
+						origin: "/stack/test.tm",
+						body:   "global-data-/stack",
```
Comment (i4ki) [inline generate/genfile/genfile_test.go:146]: top!

File: hcl/eval/eval.go [hunk 3]
```diff
@@ -75,6 +75,32 @@ func (c *Context) SetNamespace(name string, vals map[string]cty.Value) error {
...
+
+//logger.Trace().Msg("Add global evaluation namespace.")
+
+//if err := c.SetNamespace("global", globals.Attributes()); err != nil {
+//return errors.E(sm, err, "setting global namespace on eval context")
+//}
+
+//return nil
+//}
+
```
Comment (i4ki) [inline hcl/eval/eval.go:103]:
```suggestion
```

File: hcl/hcl.go [hunk 4]
```diff
@@ -415,6 +454,33 @@ func validateGenerateHCLBlock(block *hclsyntax.Block) error {
...
+	schema := &hcl.BodySchema{
+		Attributes: []hcl.AttributeSchema{
+			{
+				Name:     "content",
+				Required: true,
+			},
+		},
+	}
+
+	_, diags := block.Body.Content(schema)
```
Comment (i4ki) [inline hcl/hcl.go:477]: this gives me shivers :joy: 
When refactoring for the config loader I'd like to property parse everything ourselves.

File: test/errors/errors.go [hunk 5]
```diff
@@ -84,6 +84,9 @@ func AssertAsErrorsList(t *testing.T, err error) {
 }
 
 func errstr(err error) string {
+	if err == nil {
```
Comment (i4ki) [inline test/errors/errors.go:87]: just curious, when does this happen?

**Review by i4ki — DISMISSED** [review]
lgtm

File: generate/genfile/genfile.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+func (f File) Origin() string {
+	return f.origin
+}
+
+// GeneratedFiles returns all generated files, mapping the file path to
+// the file description. The path is absolute relative to the project root.
+func (s StackFiles) GeneratedFiles() map[string]File {
+	cp := map[string]File{}
+	for k, v := range s.files {
+		cp[k] = v
```
[the remark this answers is not in the document]
Comment (katcipis) [reply generate/genfile/genfile.go:71]: Only required to ensure that StackFiles is immutable. But it is debatable if we should ensure, Im usually biased towards ensuring =P

File: hcl/hcl.go (hunk 4 above)
Comment (katcipis) [reply hcl/hcl.go:477]: I was not that aware that the schema checks from the HCL lib were terrible, but I believe you :laughing:

File: test/errors/errors.go (hunk 5 above)
Comment (katcipis) [reply test/errors/errors.go:87]: Here: https://github.com/mineiros-io/terramate/pull/330/files#diff-e417d97c87669888e22d2588da300c8573944c4b3380df8926361027a16944adR57

One of the errors was nil, then errors.Is returned false but then I got a panic when trying to print the error with errstr.

**Review by i4ki — APPROVED** [review]

**Pull request merged**